feat(runtime): automatically insert `key` attrs during compilation

alicewriteswrongs opened the pull request:
Putting up the code for this now so we can start looking at it!

## What is the current behavior?

We have a number of issues in the backlog relating to shortcomings in how children are reconciled across virtual DOM re-renders at present. For example, #1968 is caused by an interaction between a failure to correctly recognize the same child across a re-render and another issue with slot content relocation in a `scoped` context.

This implements an approach to addressing a subset of these issues by automatically inserting `key` attributes at compile-time, making it easier at run-time for the virtual DOM to determine when two children of a virtual DOM node should be matched up with each other.


## What is the new behavior?

Now as part of transpilation Stencil will insert randomly-generated `key` attributes into JSX syntax tree nodes. This allows the more-accurate key-based child reconciliation algorithm to be used almost all the time, fixing some edge cases where previously the Stencil runtime would accidentally misplace a child, failing to, for instance, correctly match up an old and new vdom node when nodes were reordered.

- fixes #1968 
- fixes #5263 

## Does this introduce a breaking change?

I don't believe it does!

- [ ] Yes
- [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Testing

I have added some tests which exercise the change and also ensured that all of our existing tests are passing without issue.

I also built and ran this in Framework, where all of the unit tests are passing and some smoke testing of components (the datetime in particular) doesn't reveal any potential problems.

But! For testing this I think it would be a good idea to check out the reproduction case from   #1968. Then:

- confirm you can reproduce the issue with `@stencil/core@latest`
- checkout this branch
- `npm run build && npm pack`
- install the tarball in the reproduction case
- confirm that the issue is fixed

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts:
```diff
@@ -0,0 +1,88 @@
+  return transpileModule(code, null, null, [performAutomaticKeyInsertion]);
+}
+
+describe('convert-decorators', () => {
```
these aren't real tests right now, more just for debugging the impl

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/index.ts:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  if (jsxElement.tagName.getText() === '') {
+    // TODO do we need this check?
```
I don't know if this is necessary, leaving it in for now so I remember to check later

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/index.ts:
```diff
@@ -0,0 +1,158 @@
+  ]);
+
+  if (jsxElement.tagName.getText() === 'rect' || jsxElement.tagName.getText() === 'svg') {
+    // TODO diagnose the issue with svg tags
```
svg attrs are _sometimes_ being dropped without this check, I'm not sure why but I'm going to leave this for now so I remember to investigate fully and then fix that issue

alicewriteswrongs commented on src/runtime/test/render-vdom.spec.tsx:
```diff
@@ -337,7 +337,7 @@ describe('render-vdom', () => {
         vdomClass: false,
         vdomStyle: false,
-        vdomKey: false,
+        vdomKey: true,
```
these tests for `vdomKey` just all changed because now every build has `key` attrs in it

alicewriteswrongs commented on test/karma/test-app/scoped-conditional/cmp.tsx:
```diff
@@ -0,0 +1,31 @@
+      </div>
+    );
+  }
+}
```
this is from a commit created by @rwaskiewicz which reproduces #1968 as a karma test.

alicewriteswrongs replied:
(ditto for the other stuff in this directory)

before adding the transformer these tests were failing, now they pass!

alicewriteswrongs reviewed (commented):
noting a few things

alicewriteswrongs commented:
I just published `@stencil/core@4.8.1-dev.1701985394.df7b575` from this branch, so that can be used to assess whether this addresses open issues

christian-bromann reviewed (approved):
LGTM 👍

rwaskiewicz commented on test/karma/test-app/slot-reorder/karma.spec.ts:
```diff
@@ -72,7 +75,7 @@ describe('slot-reorder', () => {
       expect(r.children[0].getAttribute('name')).toBe('slot-b');
       expect(r.children[1].textContent.trim()).toBe('fallback default');
-      expect(r.children[1].hasAttribute('hidden')).toBe(false);
+      expect(r.children[1].hasAttribute('hidden')).toBe(true);
```
I'm not sure I fully understand this - the implication of this code change to me is that the `'hidden'` attribute's presence/value would be changing from today's existing behavior for scoped components - can you help me understand if this is the case / why we need to change this value?

alicewriteswrongs replied:
I believe what's going on is that previously this element was being destroyed and re-created so that it didn't have the hidden attribute anymore, now it is preserved across re-renders so although the _value_ of `el.getAttribute('hidden')` is I believe consistent and what it's supposed to be we see a difference in whether or not the attribute has been set on the element. Does that make sense?

rwaskiewicz replied:
I'm not sure, I'm still not totally clear as to how this is a result of automatic key insertion. Who/what was doing the recreation before, and who's doing it now? I suppose what I'm trying to understand here is any differences in our render cycle that might be a result of this change, and this question is just a small part of that.

alicewriteswrongs replied:
I realized that this is another situation where I think we want to bail before adding keys to the JSX. This is the component that's being tested in this file:

https://github.com/ionic-team/stencil/blob/148e8cd45d96e9a331a3aac307e49a2b18fef0e0/test/karma/test-app/slot-reorder/cmp.tsx#L1-L40

In particular, note the `render()` method has two returns, each of which wrap some slot stuff with `<div>` tags, like so:

```tsx
export class SlotReorder {
  @Prop() reordered = false;

  render() {
    if (this.reordered) {
      return (
        <div class="reordered">
          { ... }
        </div>
      );
    }

    return (
      <div>
        { ... }
      </div>
    );
  }
}
```

in this case we can't safely add `key` attrs to those divs, because doing so will cause the runtime to not match them up across renders. In particular, our implementation of `isSameVnode` assumes that _if keys are set_ on two vnodes then those are authoritative:

https://github.com/ionic-team/stencil/blob/148e8cd45d96e9a331a3aac307e49a2b18fef0e0/src/runtime/vdom/vdom-render.ts#L586-L604

The transformed version of the component in question here would look something like this, with keys inserted:

```tsx
export class SlotReorder {
  @Prop() reordered = false;

  render() {
    if (this.reordered) {
      return (
        <div class="reordered" key="asdfasdf1234">
          { ... }
        </div>
      );
    }

    return (
      <div key="hjklhjkl5678">
        { ... }
      </div>
    );
  }
}
```

because the keys set on those syntax nodes are _not_ equal, when the vdom assesses whether or not these nodes are equal it will see that they _have_ `key` attrs and, in that case, it will say "ok, are these equal? no? then these are not the same vnode". This will cause it to then re-render the whole thing, leading to some undesirable behavior where the `<div>` actually _should_ be kept around between re-renders so that its children can be diffed as well, rather than always being re-rendered on each render cycle.

I thought about this a bit, and I think that there's just not really a safe way that we can add keys in a case like this where there are multiple returns from the component's `render()` method. Or, at the least, I don't think we could safely do so without adding some more fancy static analysis that I dont _think_ we want to take on. Like adding keys would be safe here I think:

```tsx
export class ImageOrText {
  @Prop() showImage = false;

  render() {
    if (this.showImage) {
      return (
        <img src="..." />
      );
    }

    return (
      <span>text instead!</span>
    );
  }
}
```

because the two tags are different the vdom would already have to re-render the elements when `this.showImage` changes, so we wouldn't run into problems adding keys to the `img` and `span` tags in the JSX. But! I don't think we probably want to bite off trying to figure out when it's safe to do this. So instead I think we need to not add keys when:

1. there are two returns in a component's `render` method and
2. a component's `render` method returns a ternary

Does that all make sense? I've added some further constraints to when we transform JSX nodes and it makes this test in pass without the changes I have in this file right now.

rwaskiewicz commented:
Ahh - two more things:
1. I think we should mark this in a minor release of Stencil/mark it as a feature. Can we update the PR name/commit messages?
2. Can we add some documentation around this behavior for the Stencil Site?

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/utils.ts:
```diff
@@ -0,0 +1,36 @@
+ * @param jsxElement a typescript JSX syntax tree node which needs a key
+ * @returns a computed unique key for that element
+ */
+export function deriveJSXKey(jsxElement: ts.JsxOpeningElement | ts.JsxSelfClosingElement): string {
+  const hash = createHash('sha1')
+    .update(`${incrementer.next().value}__${jsxElement.tagName}__${jsxElement.pos}_${jsxElement.end}`)
+    .digest('hex')
+    .toLowerCase();
+  return hash;
+}
```
I have some numbers on the JIRA ticket (STENCIL-893) and the tl;dr is that there's no discernible or significant performance difference when building Framework

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/index.ts:
```diff
@@ -0,0 +1,136 @@
+    if (ts.isClassDeclaration(node)) {
+      return ts.visitEachChild(node, findRenderMethodVisitor, transformCtx);
+    } else {
+      return ts.visitEachChild(node, findClassDeclVisitor, transformCtx);
```
possibly not, just sort of a completeness / correctness thing I guess

but yeah since a class decl will always be a top-level syntax node with a `SourceFile` we can probably just return here, which will cut out some tree-traversal we'd do otherwise

alicewriteswrongs replied:
I actually found out that we do support non-top-level Stencil components in at least one important case - our own tests files, like here for instance:

https://github.com/ionic-team/stencil/blob/cea23c4a306d05a41f9ca137e9168039276ba30a/src/runtime/test/render-vdom.spec.tsx#L8-L28

I think if we want to continue to support those we'll need to keep traversing the whole tree, unless we want to bite off some more added complexity

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/index.ts:
```diff
@@ -0,0 +1,136 @@
+   */
+  function findRenderMethodVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isMethodDeclaration(node) && node.name.getText() === 'render') {
+      return ts.visitEachChild(node, jsxElementVisitor, transformCtx);
```
mm that's a good point, we probably would want to do something like that anyway to avoid transforming someone's weird in-a-stencil-project React components or something

I'll pull those lines form `convert-decorators` out into a helper (`classDeclIsStencilComponent` or something)

alicewriteswrongs replied:
So! I wasn't able to use the code from `convert-decorators` here because the relevant syntax nodes have already been removed from the component class node when they're 'static-ized' in `convertDecoratorsToStatic`, _however_ I was able to add a check which looks for the `is` property, as we already do here in `parseStaticComponentMeta`:

https://github.com/ionic-team/stencil/blob/cea23c4a306d05a41f9ca137e9168039276ba30a/src/compiler/transformers/static-to-meta/component.ts#L49-L53

this lets me determine whether or not a class is a Stencil component, and gate the transformation on that. I added a test to ensure that a non-decorated class which has a `render()` method with JSX doesn't get transformed.

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/index.ts:
```diff
@@ -0,0 +1,136 @@
+function addKeyAttr(
+  jsxElement: ts.JsxOpeningElement | ts.JsxSelfClosingElement,
+): ts.JsxOpeningElement | ts.JsxSelfClosingElement {
+  const updatedAttributes = ts.factory.createJsxAttributes([
+    ts.factory.createJsxAttribute(
+      ts.factory.createIdentifier('key'),
+      ts.factory.createStringLiteral(deriveJSXKey(jsxElement)),
+    ),
+    ...jsxElement.attributes.properties,
+  ]);
```
it won't because we splat in the existing attrs after the new one

alicewriteswrongs replied:
ah so this actually works more strangely that one might think!

I think TypeScript is perfectly fine with you doing

```tsx
<foo-bar key="first" key="second" />
```

because what it means to have multiple keys with the same name is really a runtime question

I'll add a check to ensure there's not already a property with the name `"key"` before we add

alicewriteswrongs replied:
I added that check and also a test that if you've already set a `key` attr we won't add another one

alicewriteswrongs commented on src/runtime/test/render-vdom.spec.tsx:
```diff
@@ -902,15 +902,17 @@ describe('render-vdom', () => {
 
-      const rect = root.querySelector('rect');
+      let rect = root.querySelector('rect');
       expect(rect.getAttribute('transform')).toBe(null);
 
       root.isOpen = true;
       await waitForChanges();
+      rect = root.querySelector('rect');
       expect(rect.getAttribute('transform')).toBe('rotate(45 27 27)');
 
       root.isOpen = false;
       await waitForChanges();
+      rect = root.querySelector('rect');
```
a small difference in how the elements get matched up means that where before a new `rect` element wouldn't be created on re-render it now will be, so the reference to the `rect` element created on the first render isn't to the updated element created on the second re-render

alicewriteswrongs replied:
It's possible if a stencil user is writing very detailed tests that assert that within a particular component a given HTML element is either destroyed and recreated _or_ that it's conserved across re-renders they might need to rewrite some tests. I have tested in Framework and all the tests there pass with no issues, so that's a good sign at least. Possibly we want to alert users to this? I'm not sure I would really want to classify it as a breaking change though since I think that we are fully within our rights to 'claim' the specific behavior of how Stencil's vdom produces the actual DOM that the user declaratively specifies as an implementation detail, and not itself part of the public API for building Stencil components and so on. I also haven't been able to produce counter-examples for myself which would show different user-facing behavior with this change, only 'under the hood' differences, and I think it's likely that most users wouldn't need to edit their tests with this change

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/utils.ts:
```diff
@@ -0,0 +1,36 @@
+const incrementer = (function* () {
+  let val = 0;
+  while (true) {
+    yield val++;
```
haha yeah that would be a big stencil project!

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts:
```diff
@@ -0,0 +1,125 @@
+`,
+    );
+  });
+});
```
good idea, I added a test to make sure that we don't add keys in cases where we can't assure ourselves it's safe to do so

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/automatic-key-insertion.spec.ts:
```diff
@@ -0,0 +1,253 @@
+    );
+  });
+
+  it('should not add a static key to dynamic elements', async () => {
+    jest.spyOn(keyInsertionUtils, 'deriveJSXKey').mockReturnValueOnce('once-key');
+    const t = transpile(`
+     @Component({tag: 'cmp-a'})
+     export class CmpA {
+        render() {
+          return (
+            <div>
+              {this.todos.map((todo) => (
+                <div>{ todo }</div>
+              ))}
+            </div>
+          )
+      }
+    }`);
+    expect(await formatCode(t.outputText)).toBe(
+      `export class CmpA {
+  render() {
+    return h(
+      'div',
+      { key: 'once-key' },
+      this.todos.map((todo) => h('div', null, todo)),
+    );
+  }
+  static get is() {
+    return 'cmp-a';
+  }
+}
+`,
+    );
+  });
```
this is a test to ensure that we don't add static keys to dynamic generated elements. I did a bit of a conservative cut-off for these - we don't attempt to transform any JSX nodes which are found within a JSX expression, so with (contrived) code like

```ts
@Component({ tag: 'cmp-a' })
class CmpA {
  render() {
    return <div>{ (() => <div>inner</div>)() }</div>
  }
}
```

we won't transform the 'inner' div in the IIFE because, since it will be generated at runtime, we could run into a lot of cases where we can't _really_ know too much about what's going on there and when it would not be safe to add keys.

alicewriteswrongs reviewed (commented):
just noting a few things

alicewriteswrongs commented:
@rwaskiewicz I believe this is ready for another look! I've addressed your comments and added a few more tests to cover some more edge cases, lmk what you think!

alicewriteswrongs commented on src/compiler/transformers/automatic-key-insertion/index.ts:
```diff
@@ -0,0 +1,197 @@
+      if (ts.isStringLiteral(expression) || ts.isNumericLiteral(expression)) {
+        return expression.text;
+      }
+    default:
```
yeah that's a good idea 👍

rwaskiewicz reviewed (changes requested):
Overall looks good!

I have two asks before we can merge this:
1. I had one change I'd like to make to the code in its current state (comment made as a part of this review)
2. Did we ever push the PR up containing docs around this functionality?

alicewriteswrongs commented:
@rwaskiewicz I removed those changes, thanks for checking if they were still necessary!

alicewriteswrongs commented:
also I opened a documentation here: https://github.com/ionic-team/stencil-site/pull/1330

christian-bromann reviewed (approved):
Struggled a bit to reproduce the problem but with the help of @alicewriteswrongs reproducible example I was able that the issue is resolved with this patch 👍
